nomad/deployment_endpoint.go

```diff
@@ -407,17 +407,26 @@ func (d *Deployment) List(args *structs.DeploymentListRequest, reply *structs.De
 			// Capture all the deployments
 			var err error
 			var iter memdb.ResultIterator
+			namespace := args.RequestNamespace()
+
 			if prefix := args.QueryOptions.Prefix; prefix != "" {
-				iter, err = store.DeploymentsByIDPrefix(ws, args.RequestNamespace(), prefix)
-			} else if args.RequestNamespace() == structs.AllNamespacesSentinel {
-				iter, err = store.Deployments(ws)
+				iter, err = store.DeploymentsByIDPrefix(ws, namespace, prefix)
+			} else if namespace != structs.AllNamespacesSentinel {
+				iter, err = store.DeploymentsByNamespaceOrdered(ws, namespace, args.OrderAscending)
 			} else {
-				iter, err = store.DeploymentsByNamespace(ws, args.RequestNamespace())
+				iter, err = store.Deployments(ws, args.OrderAscending)
 			}
 			if err != nil {
 				return err
 			}
 
+			iter = memdb.NewFilterIterator(iter, func(raw interface{}) bool {
+				if deployment := raw.(*structs.Deployment); deployment != nil {
+					return args.ShouldBeFiltered(deployment)
+				}
+				return false
+			})
```

Comment:
Because we now rely on a `_prefix` iterator over namespaces, we need to filter out any deployments that erroneously match the prefix. (e.g. want `foo`, but also matches `fooo`).

Reply:
Maybe it would be worth having a test for this? Just to make sure the behaviour or `ShouldBeFiltered` changes unexpectedly.